### DISCUSSION TITLE
Submit Forms to Keycloak

### DISCUSSION THREAD

**Discussion by chmcquee** [keycloakify/keycloakify#247] [Q&A] [answered]
I'm back with a question about the submission of the forms. Currently, I'm using Formik and Yup for the forms. Now my question is, how do we send data (and in what way) to keycloak so it will validate what we send?

From the live demo app, I don't understand how to do this in the onSubmit function.

Many thanks for considering my request.
Chris

**Comment by garronej** [comment, marked as answer]
Hello @chmcquee,  
There is nothing specific to Keycloak to know.  
The form is a classic HTML form.  
I'd suggest you take a step back and take the time to study https://developer.mozilla.org/en-US/docs/Learn/Forms  

Best regards,

**Reply by chmcquee** [reply]
Thanks for the link,
was wondering how the request is made to Keycloak and what URL to target.

**Reply by chmcquee** [reply]
My form now looks like this ` <form
                  id="kc-form-login"
                  onSubmit={handleSubmit}
                  action={url.loginAction}
                  method="post"
                >`
Now I wanted to test it in our app and even with a valid username and password I get no response when clicking the login button:
`   <input
                      tabIndex={4}
                      className={"submit-button"}
                      name="login"
                      id="kc-login"
                      type="submit"
                      value={msgStr("doLogIn")}
                      disabled={!isValid || isSubmitting}
                    />`
Is there anything in the keycloak admin I need to set to get a redirect after the login or error if login failed?
Struggling to get this implemented.
Thanks again.

**Reply by garronej** [reply]
Have you checked this? https://docs.keycloakify.dev/realtime-input-validation

**Reply by chmcquee** [reply]
Yes, I have, just I don't have the option User Profile Enabled in the Realm Settings we have to use.

**Reply by garronej** [reply]
You have to enable it manually when launching Keycloak:  
NOTE: User profile is only available in Keycloak 15 and it's a beta feature that [needs to be enabled when launching keycloak](https://github.com/InseeFrLab/keycloakify/blob/59f106bf9e210b63b190826da2bf5f75fc8b7644/src/bin/build-keycloak-theme/build-keycloak-theme.ts#L116-L117) and [enabled in the console](https://user-images.githubusercontent.com/6702424/136874428-b071d614-c7f7-440d-9b2e-670faadc0871.png).  

If you can't enable this feature for some reason, you're out of luck.

**Reply by chmcquee** [reply]
@garronej  ok understood, but is this the only way to validate password and email AND post the values to keycloak to verify?

I mean is there a way to send a Promise to keycloak with the data? 
Or can we see what the url.loginAction is submitting and what happens on failure/ success?

**Reply by chmcquee** [reply]
Maybe to clarify: We're using our custom setup for the registration but sending the data to keycloak. So I need verification for the login and feedback if the credentials are false.

**Reply by chmcquee** [reply]
@garronej  I've now disabled Formik and just want to submit the basic form. How can I determine the links to redirect to after successful login? Whenever I login the page just reloads....

**Reply by garronej** [reply]
Again this is out of the Keycloakify scope.  
See: https://wjw465150.gitbooks.io/keycloak-documentation/content/securing_apps/topics/oidc/javascript-adapter.html  
And this is how I personaly do it: https://github.com/InseeFrLab/onyxia-web/blob/793b809908a72d988fd124307d909f25801ea17e/src/core/adapters/keycloakOidcClient.ts#L33-L46

**Reply by chmcquee** [reply]
hi @garronej,
thanks so much for your patience. In the end we found the issue. Our submit was an input instead of a button. The input also had the name "login" which was being sent to keycloak and somehow was causing the session to not be created. 
In the end this is what worked for us ` <button
                      tabIndex={4}
                      className={"submit-button"}
                      id="kc-login"
                      type="submit"
                      disabled={
                        !isValid ||
                        isSubmitting ||
                        errors.password !== undefined ||
                        errors.username !== undefined
                      }
                    >
                      {msgStr("doLogIn")}
                    </button>`
Wanted to share this, as it may help somebody in the future :)

**Reply by garronej** [reply]
Thanks for your feedback!

**Comment by garronej** [comment]
Aslo @chmcquee,  
I recomend against using Formik and Yup.  
It's good to have client side validation of input client side for the UX but it must also be checked on the backend otherwise your system can easyly be taken advantage of (an attaker can manually submit whatever data he wants).  

On top of that you probably wants to define your validation criteria from the Keycloak admin instead of hard coding it in the them code.  

The recomended approach (featured out of the box) for client side validatation and back end validation is this one: https://docs.keycloakify.dev/realtime-input-validation

**Reply by chmcquee** [reply]
I think its very important to validate both on the client side AND backend. Both should be necessary for maximum safety.

**Reply by garronej** [reply]
Yes this is precisely why I recomend checking out https://docs.keycloakify.dev/realtime-input-validation
